specifyweb/frontend/js_src/lib/components/Header/Notifications.tsx

```diff
@@ -42,10 +38,21 @@ export function Notifications({
     if (notificationCount === 0) handleClose();
   }, [notificationCount, handleClose]);
 
-  React.useEffect(() => {
-    let pullInterval = INITIAL_INTERVAL;
-    let lastFetchedTimestamp: Date | undefined;
+  let pullInterval = INITIAL_INTERVAL;
+  let lastFetchedTimestamp: Date | undefined;
+  let destructorCalled = false;
+  const timeout: NodeJS.Timeout | undefined = undefined;
+
+  const { doFetch } = useNotificationsFetch({
+    pullInterval,
+    freezeFetchPromise,
+    destructorCalled,
+    notifications: [notifications, setNotifications],
+    lastFetchedTimestamp,
+    timeout,
```

Comment:
move all of these into the hook

the hook should return just `notifications` and `doFetch` OR the hook can also take care of fetching (move the useEffect into the hook) and just return `notifications` and nothing else - you see how nicely that would move all of the notification fetching business into a hook, thus it can be tested independently of all the rendering logic (which would be in this file)

Reply:
Do you want to move the state notifications in the hook?

Reply:
Yes, the notifications state can be moved into the hook. 
I think the most important concept here is the decoupling of the logic which fetches the Notifications and the logic which renders them. 
Thus this can be done in a few ways (of which @maxxxxxdlp has mentioned): 
- the hook can return a (probably asynchronous) function (like `doFetch`) which returns the fetched notifications
  - This gives the most flexibility to the consumer of the hook, but that flexibility comes at a price and requires more work to be done _by_ the consumer
  
- the hook can initially fetch the notifications and return the fetched notifications and a way to re-fetch them
  - A more "laissez-faire" approach to the hook. The concept here might be to allow the consumer immediate access to the fetched notifications, but provide  control over when future fetching is needed
  
- the hook can only return the fetched notifications and handle all fetching logic
  - _All_ of the fetching logic should be included in the hook. This is the easiest for a consumer to use, but does sacrifice a little in the way of flexibility when compared to the other methods. (Consistency and rigidity probably also means this is the least bug-prone of the cases, at least once all of the initial bugs are ironed out)